ui/src/components/Settings/ApiKeySettings.tsx

```diff
@@ -0,0 +1,231 @@
+
+const GENERATE_MESSAGE = (
+  <FormattedMessage
+    id="settings.api_key.generate"
+    defaultMessage="Generate API key"
+  />
+);
+
+const REGENERATE_MESSAGE = (
+  <FormattedMessage
+    id="settings.api_key.regenerate"
+    defaultMessage="Regenerate API key"
+  />
+);
+
+type State = {
+  state:
+    | 'START'
+    | 'REGENERATE_CONFIRMATION'
+    | 'REGENERATE_LOADING'
+    | 'REGENERATE_DONE'
+    | 'GENERATE_LOADING'
+    | 'GENERATE_DONE';
+  apiKey?: string;
+};
+
+type RegenerateConfirmAction = { type: 'REGENERATE_CONFIRM' };
+type RegenerateStartAction = { type: 'REGENERATE_START' };
+type RegenerateEndAction = { type: 'REGENERATE_END'; apiKey: string };
+type GenerateStartAction = { type: 'GENERATE_START' };
+type GenerateEndAction = { type: 'GENERATE_END'; apiKey: string };
+type CloseAction = { type: 'CLOSE' };
+
+type Action =
+  | RegenerateConfirmAction
+  | RegenerateStartAction
+  | RegenerateEndAction
+  | GenerateStartAction
+  | GenerateEndAction
+  | CloseAction;
+
+const STATE_TRANSITIONS: Record<Action['type'], State['state']> = {
+  REGENERATE_CONFIRM: 'REGENERATE_CONFIRMATION',
+  REGENERATE_START: 'REGENERATE_LOADING',
+  REGENERATE_END: 'REGENERATE_DONE',
+  GENERATE_START: 'GENERATE_LOADING',
+  GENERATE_END: 'GENERATE_DONE',
+  CLOSE: 'START',
+};
+
+function reducer(current: State, action: Action): State {
+  const next = {
+    state: STATE_TRANSITIONS[action.type],
+    apiKey: current.apiKey,
+  };
+
+  if (action.type === 'REGENERATE_END' || action.type === 'GENERATE_END') {
+    next.apiKey = action.apiKey;
+  }
+
+  if (action.type === 'CLOSE') {
+    next.apiKey = undefined;
+  }
+
+  return next;
+}
+
+export default function ApiKeySettings() {
+  const reduxDispatch = useDispatch<any>();
+
+  const role = useSelector(selectCurrentRole);
+  const [{ state, apiKey }, dispatch] = useReducer(reducer, { state: 'START' });
+
+  const generateApiKey = async () => {
+    const { data } = await reduxDispatch(generateApiKeyAction(role));
+    return data.api_key;
+  };
+
+  const onGenerate = async () => {
+    dispatch({ type: 'GENERATE_START' });
+    const apiKey = await generateApiKey();
+    dispatch({ type: 'GENERATE_END', apiKey });
+  };
+
+  const onRegenerate = () => {
+    dispatch({ type: 'REGENERATE_CONFIRM' });
+  };
+
+  const onConfirm = async () => {
+    dispatch({ type: 'REGENERATE_START' });
+    const apiKey = await generateApiKey();
+    dispatch({ type: 'REGENERATE_END', apiKey });
+  };
+
+  const onClose = async () => {
+    dispatch({ type: 'CLOSE' });
+  };
+
+  const expiresAt =
+    role.api_key_expires_at &&
+    convertUTCDateToLocalDate(new Date(role.api_key_expires_at));
+
+  return (
+    <>
+      <Card elevation={1}>
+        <h4 className="bp4-heading">
+          <FormattedMessage
+            id="settings.api_key.title"
+            defaultMessage="API key"
+          />
+        </h4>
+
+        <p>
+          {expiresAt && (
```

Comment:
This currently only handles API keys that will expire in the future. There should be a separate case that handles API keys that have expired.

**Case 1:** User doesn’t have an API key:

> You need an API key in order to access the Aleph API. Generate an API key to get started."

**Case 2:** User has an API key that hasn’t yet expired:

> Your API key will expire on {date}. You can regenerate your API key at any time. When you regenerate your API key, your old key will stop working.

**Case 3:** User has an expired API key.

> Your API key expired on {date}. Regenerate your API key to continue using the Aleph API.